### PR TITLE
additions to `azurerm_storage_account`

### LINKED ISSUE #15: azurerm - storage account resource has no support for custom domains
_This issue was originally opened by @tasquith as hashicorp/terraform#7800. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

I've been looking at the storage account resource and I can't seem to find a way to add a custom domain name, which is supported by the UI. 

I've had a look at the API reference, and have found that the required functionality is there (customDomain and useSubDomainName parameters). Are there any plans to implement this?

https://msdn.microsoft.com/en-us/library/azure/mt163564.aspx

Thanks,
Tom

### LINKED ISSUE #80: azurerm_storage_account: crash when importing resource with file encryption enabled
_This issue was originally opened by @lewinski as hashicorp/terraform#14160. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

### Terraform Version
Terraform v0.9.3 (Homebrew)

### Affected Resource(s)
- azurerm_storage_account

### Terraform Configuration Files
N/A

### Panic Output
https://gist.github.com/lewinski/9cc12d7a6a0698266aebaf50040baf55

### Expected Behavior
Terraform should have imported the resource.

### Actual Behavior
Terraform crashed.

### Steps to Reproduce
1. Create an Azure Storage account and enable "file" encryption (probably also triggers with queue and table encryption). Example: `az storage account create --location="westus" --name="tfbugtestfileenc" --resource-group="tfbugtest" --kind="Storage" --sku="Standard_GRS" --access-tier="Hot" --encryption="file"`
2. Try to import the resource using Terraform: `terraform import azurerm_storage_account.importbug /subscriptions/.../resourceGroups/tfbugtest/providers/Microsoft.Storage/storageAccounts/tfbugtestfileenc`
3. Crash.

### Important Factoids
Terraform import works fine when 1) no encryption is enabled, or 2) blob encryption is enabled.

The azurerm_storage_account resource does not actually support specifying file, queue, or table encryption. This might be necessary to take care of instead of fixing this issue, because then you could actually use Terraform to correctly manage this setting.

### LINKED ISSUE #110: Support replacing storage_account when changing account type
_This issue was originally opened by @discordianfish as terraform-providers/terraform-provider-azure#27. The original body of the issue is below._

<hr> 

### Terraform Version
- 0.9.8

### Affected Resource(s)
- azurerm_storage_account

### Expected Behavior
Since the account type can't be updated, I expected terraform to replace the resource instead.

### Actual Behavior
What actually happened?
```
* module.workers.azurerm_storage_account.tectonic_worker: 1 error(s) occurred:

* azurerm_storage_account.tectonic_worker: Error updating Azure Storage Account type "c99c88ef": storage.AccountsClient#Update: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="StorageAccountTypeConversionNotAllowed" Message="Storage account type cannot be changed to Premium_LRS."
```

### Steps to Reproduce
1. Create a Standard_LRS storage account with terraform and `terraform apply`
2. Change the tier to Premium_LRS storage account and `terraform apply`

### PR DESCRIPTION
**Pull request by tombuildsstuff**
- Support for File Encryption. Fixes #80
- Adding additional Import tests
- Support for Custom Domains on Storage Accounts. Fixes #15
- Splitting the storage account Tier and Replication out into separate fields. [Suggested by @cchildress in #117](https://github.com/terraform-providers/terraform-provider-azurerm/pull/117#issuecomment-310184010) (which also fixes #110)

Unfortunately testing Custom Domains is harder than anticipated - since it needs a static DNS name and storage name, which need to be globally unique. For this reason I've manually confirmed this works, but haven't included an acceptance test for it at the moment.

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
State migration tests pass:

```
 $ acctests azurerm TestAzureRMStorageAccountMigrateState
=== RUN   TestAzureRMStorageAccountMigrateState
2017/09/28 19:12:45 [INFO] Found AzureRM Storage Account State v0; migrating to v1
2017/09/28 19:12:45 [DEBUG] ARM Storage Account Attributes before Migration: map[string]string{"account_type":"Standard_LRS"}
2017/09/28 19:12:45 [DEBUG] ARM Storage Account Attributes after State Migration: map[string]string{"account_tier":"Standard", "account_replication_type":"LRS", "account_type":"Standard_LRS"}
2017/09/28 19:12:45 [INFO] Found AzureRM Storage Account State v0; migrating to v1
2017/09/28 19:12:45 [DEBUG] ARM Storage Account Attributes before Migration: map[string]string{"account_type":"Premium_GRS"}
2017/09/28 19:12:45 [DEBUG] ARM Storage Account Attributes after State Migration: map[string]string{"account_type":"Premium_GRS", "account_tier":"Premium", "account_replication_type":"GRS"}
--- PASS: TestAzureRMStorageAccountMigrateState (0.00s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm	0.014s
```

Acceptance tests passed when I ran them earlier - but I'm re-running them now to confirm

**Comment by tombuildsstuff** [comment]
screen shot 2017-09-28 at 19 21 42

File: website/docs/r/storage_account.html.markdown [hunk 1]
```diff
@@ -49,11 +49,9 @@ The following arguments are supported:
     and `BlobStorage`. Changing this forces a new resource to be created. Defaults
     to `Storage`.
 
-* `account_type` - (Required) Defines the type of storage account to be
-    created. Valid options are `Standard_LRS`, `Standard_ZRS`, `Standard_GRS`,
-    `Standard_RAGRS`, `Premium_LRS`. Changing this is sometimes valid - see the Azure
-    documentation for more information on which types of accounts can be converted
-    into other types.
+* `account_tier` - (Required) Defines the Tier to use for this storage account. Valid options are `Standard` and `Premium`. Changing this forces a new resource to be created
```
Comment (vancluever) [inline website/docs/r/storage_account.html.markdown]: Wrap the markdown at 80 characters?

**Review by vancluever — APPROVED** [review]
LGTM with a minor nitpick about the markdown wrapping :+1:

File: azurerm/resource_arm_storage_account.go [hunk 2]
```diff
@@ -376,60 +466,80 @@ func resourceArmStorageAccountRead(d *schema.ResourceData, meta interface{}) err
...
+		d.Set("secondary_location", props.SecondaryLocation)
+
+		if endpoints := props.PrimaryEndpoints; endpoints != nil {
+			d.Set("primary_blob_endpoint", endpoints.Blob)
+			d.Set("primary_queue_endpoint", endpoints.Queue)
+			d.Set("primary_table_endpoint", endpoints.Table)
+			d.Set("primary_file_endpoint", endpoints.File)
+
+			pscs := fmt.Sprintf("DefaultEndpointsProtocol=https;BlobEndpoint=%s;AccountName=%s;AccountKey=%s",
+				*endpoints.Blob, *resp.Name, *accessKeys[0].Value)
```
Comment (mbfrahry) [inline azurerm/resource_arm_storage_account.go]: Just want to confirm that we are guaranteed to have 2 keys

**Review by mbfrahry — APPROVED** [review]
LGTM with a minor comment

File: azurerm/resource_arm_storage_account.go (hunk 2 above)
Comment (tombuildsstuff) [reply azurerm/resource_arm_storage_account.go]: Yep, afaik

File: website/docs/r/storage_account.html.markdown (hunk 1 above)
Comment (tombuildsstuff) [reply website/docs/r/storage_account.html.markdown]: Taking a look into this - we've got a ton of cases which have gone beyond this limit - so I'm going to leave this as-is for now and we'll do a provider wide cleanup in a bit

Comment (vancluever) [reply website/docs/r/storage_account.html.markdown]: I know that feeling :wink: :+1:

**Pull request merged**